### PR TITLE
add Reservoir TVL on Ethereum

### PR DESCRIPTION
**Pull request by eukadish**
**NOTE**

#### Please enable "Allow edits by maintainers" while putting up the PR.

---

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
3. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
4. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
5. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
6. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Logo (High resolution, will be shown with rounded borders):
reservoirlogo

### CODE REVIEW

**Comment by llamatester** [comment]
The adapter at projects/reservoir-protocol exports TVL: 
        
 
 
    ethereum                  0
    
    total                    0

**Comment by g1nt0ki** [comment]
hi  @eukadish how are new stablecoins minted? we track value deposited into a protocol rather than token supply as tvl, in this case, only the USDC in PSM (https://etherscan.io/address/0x4809010926aec940b550d34a46a52739f996d75d) can be treated as tvl, but I am not sure how trUSD works

**Comment by g1nt0ki** [comment]
so, we have listed to kinds of RWA projects, one is tokenization like tokenized US treasury bonds, here we count the value of circulating tokens. The other is, RWA lending, in these cases, we count the value left on the contract on chain as tvl, and and part moved off chain under borrowed, feel like your protocol falls under the latter category.

**Comment by g1nt0ki** [comment]
because we are currently not track rUSD token price, and even if we do, we cant use it as tvl, please check my previous comment

**Comment by eukadish** [comment]
We would like to show the totalSupply of our tokens in a Lllama dashboard. How can I do this?

**Comment by g1nt0ki** [comment]
hi @eukadish yeah, we can list in stablecoin as well, but tvl of the protocol needs to be tracked before that

**Comment by llamatester** [comment]
The adapter at projects/reservoir-protocol exports TVL: 
        
 
 
    ethereum                  5.27 M
    
    total                    5.27 M

**Comment by eukadish** [comment]
Got it @g1nt0ki I think should be good now. I think should be good now for merge.

**Comment by g1nt0ki** [comment]
hi @eukadish feel like you missed the part where amount lent out should be counted as borrowed and only the USDC left in the contract on chain could be counted as tvl.

something like this: https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/credix/index.js

**Review by g1nt0ki — CHANGES_REQUESTED** [review]
.

**Comment by eukadish** [comment]
> hi @eukadish feel like you missed the part where amount lent out should be counted as borrowed and only the USDC left in the contract on chain could be counted as tvl.
> 
> something like this: https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/credix/index.js

That's not what we are trying to show though. We're basically trying to display our net liabilities, which in our case are always less than our net assets which are all lent out.

Maybe I should make a PR to those other repos I had before? For the other tokens.

**Comment by g1nt0ki** [comment]
yes, we track liabilities under borrowed key. Sorry, I would like to sort of tvl metric before listing stablecoin data

**Comment by llamatester** [comment]
The adapter at projects/reservoir-protocol exports TVL: 
        
 
 
    ethereum                  10.13 M
    ethereum-borrowed         0
    borrowed                  0
    
    total                    10.13 M

**Comment by g1nt0ki** [comment]
@eukadish feel like there was some miscommunication, is this 10M not yet lent out? in that case, can you update code to count USDC balance in your contracts?

**Comment by eukadish** [comment]
I don't know what you mean by lent, the 10 M is basically our liability, so yes, that is the amount that has been lent. Can we merge?

**Comment by g1nt0ki** [comment]
You keep asking me to merge but from the start, you have made zero change. if 10M is your liability, it should be counted towards your borrowed field, tvl = USDC  in your smart contract. 

If there is 10M usdc in your smart contract, please share that contract address

**Comment by eukadish** [comment]
I haven't made any changes to the request because it is correct as is. You are misunderstanding how the protocol works. The TVL of reservoir != USDC in the protocol. rUSD is backed by various assets, when rUSD is used to mint trUSD or srUSD, the rUSD is burned. The correct Reservoir TVL calc = total supply of rUSD + srUSD + trUSD. Please let me know if you have any additional questions. Happy to take this to a call or telegram chat if needed. Thanks!

**Comment by eukadish** [comment]
@g1nt0ki Kindly following up here. Please let me know if you have any outstanding questions / concerns. We really would like to solve this so we can push this live on defillama. Appreciate your help in getting this over the line.

**Comment by g1nt0ki** [comment]
> so, we have listed to kinds of RWA projects, one is tokenization like tokenized US treasury bonds, here we count the value of circulating tokens. The other is, RWA lending, in these cases, we count the value left on the contract on chain as tvl, and and part moved off chain under borrowed, feel like your protocol falls under the latter category.

.

**Comment by llamatester** [comment]
The adapter at projects/reservoir-protocol exports TVL: 
        
 
 
    ethereum                  10.06 M
    ethereum-borrowed         681.65 k
    borrowed                  681.65 k
    
    total                    10.06 M

File: projects/reservoir-protocol/index.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    const srUSDTotalSupply = await api.call({ target: srUSD, abi: 'function totalSupply() external view returns (uint256)' });
+    api.add(srUSD, srUSDTotalSupply / 1e12);
+}
+
+module.exports = {
+    ethereum: {
+        tvl,
+        borrowed: async (api) => {
+            const borrows = await api.call({
```
Comment (g1nt0ki) [inline projects/reservoir-protocol/index.js:23]: amount not lent out would be tvl, rest is borrowed. so, need to swap the keys here

**Comment by llamatester** [comment]
The adapter at projects/reservoir-protocol exports TVL: 
        
 
 
    ethereum                  9.42 M
    
    total                    9.42 M

File: projects/reservoir-protocol/index.js (hunk 1 above)
Comment (eukadish) [reply projects/reservoir-protocol/index.js]: @g1nt0ki Check now if you don't mind, changed our TVL to our full capital held by the protocol. Let me know if any issues please :)

**Comment by g1nt0ki** [comment]
@eukadish sorry for nitpicking, but this change is good, merged with excluding rUSD vault, rUSD is minted by your protocol right?

**Comment by eukadish** [comment]
Yes rUSD is minted by our protocol. Did you close the MR @g1nt0ki without merging?

**Comment by g1nt0ki** [comment]
merged it here: https://github.com/DefiLlama/DefiLlama-Adapters/commit/cb7345683b60387f05ac7d020d43135078545eb1

**Pull request closed without merging**